task/task.go

```diff
@@ -1,15 +1,122 @@
 
 import (
+	"path/filepath"
+
+	pkgerrors "github.com/pkg/errors"
 	"github.com/symflower/eval-dev-quality/evaluate/metrics"
 	"github.com/symflower/eval-dev-quality/evaluate/repository"
+	"github.com/symflower/eval-dev-quality/language"
+	"github.com/symflower/eval-dev-quality/log"
+	evalmodel "github.com/symflower/eval-dev-quality/model"
 	taskcontext "github.com/symflower/eval-dev-quality/task/context"
 )
 
+// Task defines a task to evaluate a repository.
 type Task interface {
 	// Identifier returns the task identifier.
 	Identifier() taskcontext.Identifier
 
 	// Run runs a task in a given repository.
 	Run(repository *repository.Repository) (assessments metrics.Assessments, problems []error, err error)
 }
+
+// GetTaskByTaskIdentifier returns a task based on the task identifier.
+func GetTaskByTaskIdentifier(taskIdentifier taskcontext.Identifier, logger *log.Logger, resultPath string, model evalmodel.Model, language language.Language) (task Task, err error) {
+	switch taskIdentifier {
+	case taskcontext.IdentifierWriteTests:
+		return NewTaskWriteTests(logger, resultPath, model, language), nil
+	default:
+		return nil, taskcontext.ErrTaskUnsupported
+	}
+}
+
+// TaskWriteTests holds the write test task.
+type TaskWriteTests struct {
+	// ResultPath holds the directory path where results should be written to.
+	ResultPath string
+
+	// Language holds the language for which the task should be evaluated.
+	Language language.Language
+	// Model holds the model which the task should be evaluated.
+	Model evalmodel.Model
+
+	// Logger holds the logger for this tasks.
+	Logger *log.Logger
+}
+
+var _ Task = (*TaskWriteTests)(nil)
+
+// NewTaskWriteTests returns a write test task.
+func NewTaskWriteTests(logger *log.Logger, resultPath string, model evalmodel.Model, language language.Language) (task Task) {
+	return &TaskWriteTests{
+		ResultPath: resultPath,
+		Language:   language,
+		Model:      model,
+		Logger:     logger,
+	}
+}
+
+// Identifier returns the write test task identifier.
+func (t *TaskWriteTests) Identifier() taskcontext.Identifier {
+	return taskcontext.IdentifierWriteTests
+}
+
+// TaskWriteTests generates test files for the given implementation file in a repository.
+func (t *TaskWriteTests) Run(repository *repository.Repository) (repositoryAssessment metrics.Assessments, problems []error, err error) {
+	log, logClose, err := log.WithFile(t.Logger, filepath.Join(t.ResultPath, string(t.Identifier()), evalmodel.CleanModelNameForFileSystem(t.Model.ID()), t.Language.ID(), repository.Name+".log"))
+	if err != nil {
+		return nil, nil, err
+	}
+	defer logClose()
+
+	log.Printf("Evaluating model %q on task %q using language %q and repository %q", t.Model.ID(), t.Identifier(), t.Language.ID(), repository.Name)
+	defer func() {
+		log.Printf("Evaluated model %q on task %q using language %q and repository %q: encountered %d problems: %+v", t.Model.ID(), t.Identifier(), t.Language.ID(), repository.Name, len(problems), problems)
```

Comment:
All the logging business will be the same for all tasks, so should probably create a helper and/or move into `evaluate/evaluate.go`.